presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java

```diff
@@ -771,78 +780,183 @@ public void end()
+                            columnName,
+                            fieldType.getName());
+                    builder.add(fieldType.isPrimitive() ? ParquetDummyPrimitiveConverter.converter : new ParquetDummyGroupConverter(fieldType));
+                }
+                else {
+                    if (prevFieldIndex > pair.index) {
+                        outOfOrder = true;
+                    }
+                    builder.add(createConverter(pair.type, columnName + "." + fieldType.getName(), fieldType, pair.index, true, typeManager));
+                    prevFieldIndex = pair.index;
+                }
+            }
+            // When reading parquet files, we need to pass converters and parquet types in parquet schema fields order.
+            // After getting the parquet values, we need to build presto blocks in metastore schema fields order.
+            List<BlockFieldConverter> converters = builder.build();
+            if (outOfOrder) {
+                log.info("Schema mismatch, struct field order for metastore schema for row column %s is different from that of parquet schema",
+                        columnName);
+                ImmutableList.Builder<Type> types = ImmutableList.builder();
+                converterIndexes = new int[fieldCount];
+                for (int i = 0, tmpIndex = 0; i < fieldCount; i++) {
+                    BlockFieldConverter converter = converters.get(i);
+                    converterIndexes[i] = converter.getFieldIndex();
+                    if (converterIndexes[i] < 0) {
+                        continue;
+                    }
+                    converter.setFieldIndex(tmpIndex++);
+                    types.add(converter.getType());
+                }
+                outOfOrderFieldTypes = types.build();
+            }
+            return converters;
+        }
+
+        public ParquetStructConverter(Type prestoType, String columnName, GroupType entryType, int fieldIndex, boolean useNames, TypeManager typeManager)
         {
             checkArgument(ROW.equals(prestoType.getTypeSignature().getBase()));
-            List<Type> prestoTypeParameters = prestoType.getTypeParameters();
             List<parquet.schema.Type> fieldTypes = entryType.getFields();
-            checkArgument(
-                    prestoTypeParameters.size() == fieldTypes.size(),
-                    "Schema mismatch, metastore schema for row column %s has %s fields but parquet schema has %s fields",
-                    columnName,
-                    prestoTypeParameters.size(),
-                    fieldTypes.size());
+
+            if (fieldTypes.size() == prestoType.getTypeSignature().getParameters().size()) {
+                this.converters = createConverters(columnName, prestoType.getTypeParameters(), fieldTypes, typeManager);
+            }
+            else {
+                log.info("Schema mismatch, metastore schema for row column %s has %s fields but parquet schema has %s fields",
```

Comment:
and do we need these logs at info level? Why not warn? No one will see these logs anyway unless they dig to the worker logs.